### PR TITLE
new enlarge label shuffling sampler

### PR DESCRIPTION
**Pull request by shewmingg**
a new sampler used to deal with imbalanced data, proposed from ImageNet2016 scene classification champion. Increase the accuracy rate for 0.1% in my case.

the classifier will behave badly on the data that has far less data than other classes. The function manipulates the label indices so that each class will have the same amount of training data (by randomly repeating on rarer data).

### CODE REVIEW

File: torch/utils/data/sampler.py [hunk 1]
```diff
@@ -92,6 +92,53 @@ def __iter__(self):
...
+        preidx = 0
+        for x in range(len(self.count_of_each_label)):
+            idxes = np.remainder(t.randperm(largest).numpy(), self.count_of_each_label[x]) + preidx
+            for y in idxes:
+                self.enlarged_index.append(sorted_labels[y][0])
+            
+            preidx += int(self.count_of_each_label[x])
+
+    def __iter__(self):
+        shuffle(self.enlarged_index)
```
Comment (fmassa) [inline torch/utils/data/sampler.py]: is `shuffle` defined anywhere?

File: torch/utils/data/sampler.py [hunk 2]
```diff
@@ -92,6 +92,53 @@ def __iter__(self):
...
+        return iter(self.enlarged_index)
+    
+    def __len__(self):
+        return np.amax(self.count_of_each_label)*len(self.count_of_each_label)
```
Comment (fmassa) [inline torch/utils/data/sampler.py]: Numpy is not a hard dependency of torch, and is not even imported in this file so I believe this will be an error

File: torch/utils/data/sampler.py [hunk 3]
```diff
@@ -92,6 +92,53 @@ def __iter__(self):
...
+        count_of_each_label.append(count)
+        # get the largest count among all classes. used to enlarge every class to the same amount
+        largest = int(np.amax(count_of_each_label))
+        self.count_of_each_label = count_of_each_label
+        self.enlarged_index = []
+        
+        # preidx used for find the mapping beginning of arg "sorted_labels"
+        preidx = 0
+        for x in range(len(self.count_of_each_label)):
+            idxes = np.remainder(t.randperm(largest).numpy(), self.count_of_each_label[x]) + preidx
```
Comment (fmassa) [inline torch/utils/data/sampler.py]: is `t` defined anywhere?

**Review by fmassa — CHANGES_REQUESTED** [review]
I had a quick look and I think there are a few problems with the current implementation. Some basic unit tests would have caught those problems.

Also, another possibility would be to instead of manually replicating a `indices` list, to sample from a multinomial distribution with the probabilities being a function of the frequencies for each class, and keep the size of the dataset the same as the original one?

File: torch/utils/data/sampler.py (hunk 2 above)
Comment (shewmingg) [reply torch/utils/data/sampler.py]: Thank you for the suggestions! shuffle is from random package, t is torch. Do you think by only importing the nparray it will be good? (It's my first time trying to contribute something.. sorry for these problems)

Comment (shewmingg) [reply torch/utils/data/sampler.py]: the alternative way you described also seem to be a good solution, but by doing that, maybe there is some data not trained? This implementation will iterate the whole dataset. Only some data might be trained more than once.

File: torch/utils/data/sampler.py [hunk 4]
```diff
@@ -1,4 +1,6 @@
 import torch
+import random
+import numpy as np
```
Comment (fmassa) [inline torch/utils/data/sampler.py]: numpy is not a hard dependency of pytorch, so you shouldn't do this.
2 possibilities:
- either use alternative functions instead of numpy, you can probably use torch operations
- import numpy inside your sampler class.

File: torch/utils/data/sampler.py [hunk 5]
```diff
@@ -93,6 +95,52 @@ def __len__(self):
...
+            if y == tmp:
+                count += 1
+            else:
+                if tmp != -1:
+                    count_of_each_label.append(count)
+                    count = 1
+            tmp = y
+        count_of_each_label.append(count)
+        # get the largest count among all classes. used to enlarge every class to the same amount
+        largest = int(np.amax(count_of_each_label))
```
Comment (ezyang) [inline torch/utils/data/sampler.py]: count_of_each_label is a list, just use max

File: torch/utils/data/sampler.py [hunk 6]
```diff
@@ -93,6 +95,52 @@ def __len__(self):
...
+        self.count_of_each_label = count_of_each_label
+        self.enlarged_index = []
+
+        # preidx used for find the mapping beginning of arg "sorted_labels"
+        preidx = 0
+        for x in range(len(self.count_of_each_label)):
+            idxes = np.remainder(torch.randperm(largest).numpy(), self.count_of_each_label[x]) + preidx
```
Comment (ezyang) [inline torch/utils/data/sampler.py]: Use torch.remainder

**Comment by ezyang** [comment]
Still needs tests.

**Comment by shewmingg** [comment]
What kind of test should I further make? I have tried to test it using some dummy list as well as putting it to use.

**Comment by soumith** [comment]
@shewmingg that sounds good, testing with dummy list. Just add such a test case to https://github.com/pytorch/pytorch/blob/master/test/test_dataloader.py

**Comment by weiyangfb** [comment]
still need tests
